File: examples/ruby-rspec/lib/yahtzee.rb
# frozen_string_literal: true

module Yahtzee
  module_function

  def score(dice, category)
    counts = Hash.new(0)
    dice.each { |d| counts[d] += 1 }
    total = dice.sum
    sum_of = ->(face) { counts[face] * face }
    of_a_kind = lambda do |n|
      faces = counts.select { |_face, c| c >= n }.keys
      faces.empty? ? 0 : n * faces.max
    end
    sorted_dice = dice.sort.join

    case category
    when 'ones' then sum_of.call(1)
    when 'twos' then sum_of.call(2)
    when 'threes' then sum_of.call(3)
    when 'fours' then sum_of.call(4)
    when 'fives' then sum_of.call(5)
    when 'sixes' then sum_of.call(6)
    when 'pair' then of_a_kind.call(2)
    when 'two pairs'
      pairs = counts.select { |_face, c| c >= 2 }.keys
      pairs.length >= 2 ? pairs.sum { |face| 2 * face } : 0
    when 'three of a kind' then of_a_kind.call(3)
    when 'four of a kind' then of_a_kind.call(4)
    when 'small straight' then sorted_dice == '12345' ? 15 : 0
    when 'large straight' then sorted_dice == '23456' ? 20 : 0
    when 'full house'
      counts.size == 2 && counts.values.sort == [2, 3] ? total : 0
    when 'Yahtzee' then counts.size == 1 ? 50 : 0
    when 'chance' then total
    else raise ArgumentError, "Unknown category: #{category}"
    end
  end
end

File: examples/typescript-vitest/src/yahtzee.ts
export function score(dice: ReadonlyArray<number>, category: string): number {
  const counts = new Map<number, number>()
  for (const d of dice) counts.set(d, (counts.get(d) ?? 0) + 1)
  const sum = dice.reduce((a, b) => a + b, 0)
  const sumOf = (face: number) => (counts.get(face) ?? 0) * face
  const ofAKind = (n: number) => {
    const faces = [...counts].filter(([, c]) => c >= n).map(([face]) => face)
    return faces.length > 0 ? n * Math.max(...faces) : 0
  }
  const sorted = [...dice].sort((a, b) => a - b).join('')
  switch (category) {
    case 'ones':
      return sumOf(1)
    case 'twos':
      return sumOf(2)
    case 'threes':
      return sumOf(3)
    case 'fours':
      return sumOf(4)
    case 'fives':
      return sumOf(5)
    case 'sixes':
      return sumOf(6)
    case 'pair':
      return ofAKind(2)
    case 'two pairs': {
      const pairs = [...counts].filter(([, c]) => c >= 2).map(([face]) => face)
      return pairs.length >= 2 ? pairs.reduce((a, face) => a + 2 * face, 0) : 0
    }
    case 'three of a kind':
      return ofAKind(3)
    case 'four of a kind':
      return ofAKind(4)
    case 'small straight':
      return sorted === '12345' ? 15 : 0
    case 'large straight':
      return sorted === '23456' ? 20 : 0
    case 'full house': {
      const cs = [...counts.values()].sort((a, b) => a - b)
      return counts.size === 2 && cs[0] === 2 && cs[1] === 3 ? sum : 0
    }
    case 'Yahtzee':
      return counts.size === 1 ? 50 : 0
    case 'chance':
      return sum
    default:
      throw new Error(`Unknown category: ${category}`)
  }
}
